### PR TITLE
Isofer/idata codification

### PR DESCRIPTION
**Pull request by isofer**
# Codify InferenceData Conventions and Implement Component 1 Data Wrapper

## Overview

This PR implements **Tier 1 (Pydantic Schemas)** and **Tier 2 (Data Wrapper)** from the MMM Data & Plotting Framework Architecture research. It codifies implicit InferenceData conventions using Pydantic schemas and provides a high-level wrapper API for validated data access and transformations.

## Problem Statement

The codebase has many undocumented assumptions and implicit conventions built into InferenceData structure:

- **50+ naming patterns** across variables, dimensions, and suffixes that are not documented
- **Multiple access patterns** for the same data (observed values accessible in 4 different locations)
- **No single source of truth** for "what should be in idata"
- **Fragile post-processing** - plotting functions assume specific variable names without validation
- **Scaling logic scattered** - `_original_scale` suffix convention not enforced
- **High cognitive load** - users must know that `channel_contribution_original_scale` must be manually added before plotting

This leads to cryptic error messages, breaking changes when variables are renamed, and difficulty understanding the expected InferenceData structure.

## Solution

This PR implements a two-tier codification approach:

### Tier 1: Pydantic Schemas (`idata_schema.py`)

- **`VariableSchema`**: Validates individual variables (dimensions, dtype)
- **`InferenceDataGroupSchema`**: Validates groups and their variables
- **`MMMIdataSchema`**: Complete schema for multidimensional MMM InferenceData with configuration-based creation

Schemas provide:
- **Type safety**: Catch errors at validation time, not during plotting
- **Discoverability**: Users can inspect schema to understand idata structure
- **Documentation**: Schema serves as living documentation
- **Extensibility**: New schemas for new model types without breaking existing code

### Tier 2: Data Wrapper (`idata_wrapper.py`)

- **`MMMIDataWrapper`**: High-level wrapper around InferenceData providing:
  - Validated access to observed data, channel spend, contributions
  - Automatic scaling transformations (original ↔ scaled)
  - Filtering operations (by dates, dimensions)
  - Time aggregation (weekly, monthly, quarterly, yearly, all_time)
  - Dimension aggregation (combine channels, countries, etc.)
  - Summary statistics computation

### Tier 2: Utility Functions (`idata_utils.py`)

Standalone utility functions that operate directly on `az.InferenceData`:
- `filter_idata_by_dates()` - Filter to date range
- `filter_idata_by_dims()` - Filter by dimension values
- `aggregate_idata_time()` - Aggregate over time periods
- `aggregate_idata_dims()` - Aggregate dimension values

These can be used independently without instantiating a wrapper, providing reusability and testability.

### 1. Schema-Based Validation

```python
from pymc_marketing.mmm.idata_schema import MMMIdataSchema

# Create schema based on model configuration
schema = MMMIdataSchema.from_model_config(
    custom_dims=("country",),
    has_controls=True,
    has_seasonality=True,
    time_varying=False
)

# Validate InferenceData structure
errors = schema.validate(mmm.idata)
if errors:
    print("Validation errors:", errors)
```

### 2. Data Wrapper API

```python
# Access via model property
wrapper = mmm.data

# Get observed target data (automatically finds correct source)
observed = wrapper.get_target(original_scale=True)

# Get channel spend data
channel_spend = wrapper.get_channel_spend()

# Get all contributions in original scale
contributions = wrapper.get_contributions(original_scale=True)

# Scale transformations
original = wrapper.to_original_scale("channel_contribution")
scaled = wrapper.to_scaled("channel_contribution_original_scale")

# Filtering
filtered = wrapper.filter_dates("2024-01-01", "2024-12-31")
filtered = wrapper.filter_dims(country="US", channel=["TV", "Radio"])

# Aggregation
monthly = wrapper.aggregate_time("monthly", method="sum")
total = wrapper.aggregate_time("all_time", method="sum")
combined = wrapper.aggregate_dims(
    dim="channel",
    values=["Facebook", "Instagram", "TikTok"],
    new_label="Social",
    method="sum"
)
```

### 3. Standalone Utility Functions

```python
from pymc_marketing.mmm.idata_utils import filter_idata_by_dates, aggregate_idata_time

# Use without wrapper
filtered = filter_idata_by_dates(idata, "2024-01-01", "2024-12-31")
monthly = aggregate_idata_time(filtered, "monthly", method="sum")
```

### New Files

1. **`pymc_marketing/mmm/idata_schema.py`** (436 lines)
   - `VariableSchema` class
   - `InferenceDataGroupSchema` class
   - `MMMIdataSchema` class with `from_model_config()` factory method

2. **`pymc_marketing/mmm/idata_wrapper.py`** (529 lines)
   - `MMMIDataWrapper` class with comprehensive API
   - Methods for data access, scaling, filtering, aggregation

3. **`pymc_marketing/mmm/idata_utils.py`** (306 lines)
   - Standalone utility functions for InferenceData operations
   - `filter_idata_by_dates()`, `filter_idata_by_dims()`
   - `aggregate_idata_time()`, `aggregate_idata_dims()`

4. **`tests/mmm/test_idata_schema.py`** (712 lines)
   - Comprehensive tests for schema validation
   - Tests for all variable types and configurations

5. **`tests/mmm/test_idata_wrapper.py`** (1341 lines)
   - Comprehensive tests for wrapper functionality
   - Tests for all methods and edge cases

### Modified Files

1. **`pymc_marketing/mmm/multidimensional.py`**
   - Added `data` property that returns `MMMIDataWrapper` instance
   - Automatically creates schema based on model configuration
   - Lazy initialization with caching

## Testing

- **712 lines** of schema validation tests covering:
  - Variable structure validation (dims, dtype)
  - Group validation (required/optional)
  - Configuration-based schema creation
  - Error message quality

- **1341 lines** of wrapper tests covering:
  - Data access methods (target, channel spend, contributions)
  - Scaling transformations
  - Filtering operations (dates, dimensions)
  - Aggregation operations (time, dimensions)
  - Edge cases and error handling

All tests pass and maintain backward compatibility.

## Backward Compatibility

✅ **Fully backward compatible** - This is an additive change:
- Existing code continues to work unchanged
- New API is opt-in via `mmm.data` property
- No breaking changes to existing InferenceData structure
- Schema validation is optional (can pass `schema=None`)

### Basic Usage

```python
# After fitting a model
mmm.fit(X, y)

# Access data wrapper
wrapper = mmm.data

# Get observed data
observed = wrapper.get_target()

# Get contributions
contributions = wrapper.get_contributions(original_scale=True)
```

### Advanced Usage

```python
# Filter to specific date range and aggregate monthly
monthly_data = (
    mmm.data
    .filter_dates("2024-01-01", "2024-12-31")
    .aggregate_time("monthly", method="sum")
)

# Filter by dimensions and get contributions
us_data = mmm.data.filter_dims(country="US")
us_contributions = us_data.get_contributions(original_scale=True)

# Combine multiple channels
social_data = mmm.data.aggregate_dims(
    dim="channel",
    values=["Facebook", "Instagram", "TikTok"],
    new_label="Social",
    method="sum"
)
```

### Validation

```python
# Validate InferenceData structure
schema = MMMIdataSchema.from_model_config(
    custom_dims=("country",),
    has_controls=True
)

errors = schema.validate(mmm.idata)
if errors:
    print("Validation errors:", errors)
else:
    print("✓ InferenceData structure is valid")
```

----
📚 Documentation preview 📚: https://pymc-marketing--2176.org.readthedocs.build/en/2176/

### CODE REVIEW

File: pymc_marketing/data/idata/schema.py [hunk 1]
```diff
@@ -0,0 +1,436 @@
+#   Copyright 2022 - 2026 The PyMC Labs Developers
```
Comment (cetagostini) [inline pymc_marketing/data/idata/schema.py:1]: Maybe this can exist under `pymc_marketing/data`? Only to keep data functions under a single folder.

Comment (isofer) [reply pymc_marketing/data/idata/schema.py:1]: Like a folder name pymc_marketing/data/schema ?

File: pymc_marketing/mmm/multidimensional.py [hunk 2]
```diff
@@ -622,6 +622,58 @@ def plot(self) -> MMMPlotSuite:
...
+                ),
+            )
+
+            self._data_wrapper = MMMIDataWrapper(
+                self.idata,
+                schema=schema,
+                validate_on_init=False,  # Don't validate on every access
+            )
+
+        return self._data_wrapper
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/multidimensional.py]: fixed.

File: pymc_marketing/data/idata/utils.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+        return az.InferenceData(**aggregated_groups)
+
+    # Map period to pandas offset for periodic aggregation
+    period_map = {
+        "weekly": "W",
+        "monthly": "M",
+        "quarterly": "Q",
+        "yearly": "Y",
+    }
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:192]: fixed.

File: pymc_marketing/mmm/multidimensional.py [hunk 4]
```diff
@@ -622,6 +624,53 @@ def plot(self) -> MMMPlotSuite:
...
+        schema = MMMIdataSchema.from_model_config(
+            custom_dims=self.dims if hasattr(self, "dims") and self.dims else (),
+            has_controls=self.control_columns is not None,
+            has_seasonality=self.yearly_seasonality is not None,
+            time_varying=(
+                getattr(self, "time_varying_intercept", False)
+                or getattr(self, "time_varying_media", False)
+            ),
+        )
+        return MMMIDataWrapper(self.idata, schema=schema)
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/multidimensional.py]: fixed

File: pymc_marketing/data/idata/utils.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        other_values = list(all_coords - set(values))
+
+        if other_values:
+            other_data = group.sel({dim: other_values})
+
+            # Assign new label to aggregated
+            aggregated_values = aggregated_values.expand_dims({dim: [new_label]})
+
+            # Concatenate
+            combined = xr.concat([other_data, aggregated_values], dim=dim)
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:317]: fixed

File: pymc_marketing/data/idata/mmm_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+            if var.endswith("_original_scale"):
+                # Get base variable name
+                base_name = var.replace("_original_scale", "")
+                if base_name in self.idata.posterior:
+                    return self.idata.posterior[base_name]
+
+            if var in self.idata.posterior:
+                data = self.idata.posterior[var]
+            else:
+                raise ValueError(f"Variable '{var}' not found in posterior")
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/mmm_wrapper.py:407]: fixed

File: pymc_marketing/data/idata/utils.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+            continue
+
+        # Select values to aggregate
+        selected = group.sel({dim: values})
+
+        # Aggregate
+        if method == "sum":
+            aggregated_values = selected.sum(dim=dim)
+        elif method == "mean":
+            aggregated_values = selected.mean(dim=dim)
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:302]: not gonna fix

File: pymc_marketing/data/idata/mmm_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+
+    @property
+    def dates(self) -> pd.DatetimeIndex:
+        """Get date coordinate."""
+        if hasattr(self.idata, "constant_data"):
+            return pd.DatetimeIndex(self.idata.constant_data.coords["date"].values)
+        elif hasattr(self.idata, "posterior"):
+            return pd.DatetimeIndex(self.idata.posterior.coords["date"].values)
+        else:
+            raise ValueError("Could not find date coordinate in InferenceData")
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/mmm_wrapper.py:634]: not gonna fix

File: pymc_marketing/mmm/multidimensional.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/multidimensional.py]: fixed

File: pymc_marketing/data/idata/utils.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    aggregated_groups = {}
+    for group_name in idata.groups():
+        group = getattr(idata, group_name)
+
+        if dim not in group.dims:
+            aggregated_groups[group_name] = group
+            continue
+
+        # Select values to aggregate
+        selected = group.sel({dim: values})
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:296]: wont fix

File: pymc_marketing/mmm/multidimensional.py [hunk 10]
```diff
@@ -622,6 +624,53 @@ def plot(self) -> MMMPlotSuite:
...
+            monthly = mmm.data.filter_dates("2024-01-01", "2024-12-31").aggregate_time(
+                "monthly"
+            )
+        """
+        self._validate_idata_exists()
+
+        schema = MMMIdataSchema.from_model_config(
+            custom_dims=self.dims if hasattr(self, "dims") and self.dims else (),
+            has_controls=self.control_columns is not None,
+            has_seasonality=self.yearly_seasonality is not None,
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/multidimensional.py]: fixed

File: pymc_marketing/mmm/idata_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,628 @@
...
+                "idata validation failed:\n" + "\n".join(f"  - {e}" for e in errors)
+            )
+
+    def is_valid(self) -> bool:
+        """Check if idata structure is valid."""
+        if self.schema is None:
+            return True  # No schema, assume valid
+
+        errors = self.validate()
+        return len(errors) == 0
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/mmm/idata_wrapper.py]: fixed

File: pymc_marketing/data/idata/utils.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+        if method == "sum":
+            aggregated_values = selected.sum(dim=dim)
+        elif method == "mean":
+            aggregated_values = selected.mean(dim=dim)
+        else:
+            raise ValueError(f"Unknown aggregation method: {method}")
+
+        # Get other values (not aggregated)
+        all_coords = set(group[dim].values)
+        other_values = list(all_coords - set(values))
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:308]: not gonna fix

File: pymc_marketing/data/idata/utils.py [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+    --------
+    >>> # Combine social channels into one
+    >>> combined = aggregate_idata_dims(
+    ...     idata,
+    ...     dim="channel",
+    ...     values=["Facebook", "Instagram", "TikTok"],
+    ...     new_label="Social",
+    ...     method="sum",
+    ... )
+    """
```
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:259]: won't fix

File: pymc_marketing/data/idata/schema.py (hunk 1 above)
Comment (juanitorduz) [reply pymc_marketing/data/idata/schema.py:1]: Maybe https://github.com/pymc-labs/pymc-marketing/tree/main/pymc_marketing/data /idata/schema.py

File: pymc_marketing/data/idata/utils.py [hunk 14]
```diff
@@ -0,0 +1,324 @@
+#   Copyright 2022 - 2026 The PyMC Labs Developers
```
Comment (juanitorduz) [inline pymc_marketing/data/idata/utils.py:1]: same here Maybe https://github.com/pymc-labs/pymc-marketing/tree/main/pymc_marketing/data /idata/utils.py

File: pymc_marketing/data/idata/mmm_wrapper.py [hunk 15]
```diff
@@ -0,0 +1,623 @@
+#   Copyright 2022 - 2026 The PyMC Labs Developers
```
Comment (juanitorduz) [inline pymc_marketing/data/idata/mmm_wrapper.py:1]: Same here Maybe https://github.com/pymc-labs/pymc-marketing/tree/main/pymc_marketing/data /idata/wrapper.py

Comment (juanitorduz) [reply pymc_marketing/data/idata/mmm_wrapper.py:1]: or even better mmm_wrapper?

File: pymc_marketing/data/idata/mmm_wrapper.py [hunk 16]
```diff
@@ -0,0 +1,623 @@
...
+                "Expected 'target_data' variable in idata.constant_data."
+            )
+
+        data = self.idata.constant_data.target_data
+        if original_scale:
+            return data
+        else:
+            # Scale down using target_scale
+            target_scale = self.idata.constant_data.target_scale
+            return data / target_scale
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_marketing/data/idata/mmm_wrapper.py:124]: +1 This is one of these weird edge cases to consider.

File: pymc_marketing/data/idata/mmm_wrapper.py [hunk 17]
```diff
@@ -0,0 +1,620 @@
...
+        >>> scaled = wrapper.to_scaled("channel_contribution")
+        >>>
+        >>> # Convert DataArray from original to scaled space
+        >>> original_data = posterior["mu"] * constant_data.target_scale
+        >>> scaled = wrapper.to_scaled(original_data)
+        """
+        if isinstance(var, str):
+            if var.endswith("_original_scale"):
+                # Get base variable name (which is in scaled space)
+                base_name = var.replace("_original_scale", "")
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pymc_marketing/data/idata/mmm_wrapper.py:398]: +1

[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/mmm_wrapper.py:398]: not fixing this edge case

File: pymc_marketing/data/idata/utils.py (hunk 14 above)
Comment (isofer) [reply pymc_marketing/data/idata/utils.py:1]: I moved the files.

File: pymc_marketing/data/idata/schema.py (hunk 1 above)
Comment (isofer) [reply pymc_marketing/data/idata/schema.py:1]: I've moved the files. see comment below

File: pymc_marketing/data/idata/mmm_wrapper.py (hunk 15 above)
Comment (isofer) [reply pymc_marketing/data/idata/mmm_wrapper.py:1]: moved and renamed to mmm_wrapper

File: pymc_marketing/data/idata/mmm_wrapper.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (isofer) [reply pymc_marketing/data/idata/mmm_wrapper.py:124]: fixed

**Review by juanitorduz — APPROVED** [review]
amazing @isofer !
